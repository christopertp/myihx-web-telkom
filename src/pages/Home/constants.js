export const BILL_FETCHED = 'Home/bill-fetched';
export const CARD_FETCHED = 'Home/card-fetched';
export const DATA_FETCHED = 'Home/data-fetched';
export const LOADING = 'Home/loading';
export const NOTIF_FETCHED = 'Home/notif-fetched';
export const PROFILE_FETCHED = 'Home/profile-fetched';
