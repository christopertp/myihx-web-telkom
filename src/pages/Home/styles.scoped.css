.finder {
  background-color: #ffffff;
  padding-top: 3rem;
}

.user {
  display: inline-grid;
  grid-template-columns: auto auto;
  grid-template-rows: auto auto auto;
  margin-bottom: 1rem;
}

.user > figure {
  background-color: #f2f2f2;
  background-repeat: no-repeat;
  background-size: cover;
  border: 0.125rem solid #e5e5e5;
  border-radius: 50%;
  grid-row: 1/4;
  height: 4rem;
  margin: auto 1rem auto 0;
  width: 4rem;
}

.user > span:nth-child(2) {
  color: rgba(255, 255, 255, 0.5);
  font-size: 0.8rem;
  font-weight: 800;
  text-transform: uppercase;
}

.user > span:nth-child(3) {
  color: #ffffff;
  font-size: 1.25rem;
  font-weight: 800;
}

.user > span:nth-child(4) {
  color: #ffffff;
  font-size: 0.9rem;
  font-weight: 400;
}

.welcome > small {
  color: #ffffff;
  display: block;
  font-size: 0.875rem;
  font-weight: 400;
}

.welcome > p {
  color: #ffffff;
  font-size: 1.25rem;
  font-weight: 800;
  margin-bottom: 1.5rem;
}

.upgrade-banner {
  background: url(/assets/bnr_upgrade.jpg);
  background-color: #f2f2f2;
  background-size: cover;
  background-repeat: no-repeat;
  background-position: right;
  margin: -4rem var(--neg-side-margin) 0;
  padding: 8rem var(--side-margin);
}

.upgrade-banner > h2 {
  font-size: 2.5rem;
  color: #ffffff;
  margin-bottom: 0.25rem;
}

.upgrade-banner > p { 
  color: rgba(255, 255, 255, 0.75);
  font-size: 1.4rem;
}

.upgrade-banner > a {
  display: inline-block;
  margin-top: 0.5rem;
  border-radius: 0.3rem;
}

@media only screen and (max-width: 48em) {
  .finder {
    background-color: #f2f2f2;
    padding-top: 1rem;
  }
  .user {
    grid-template-columns: auto auto;
    grid-template-rows: auto auto;
  }  
  .user > figure {
    grid-row: 1/3;
    height: 2.5rem;
    margin: auto 0.625rem auto 0;
    width: 2.5rem;
  }
  .user > span:nth-child(2) { display: none; }
  .user > span:nth-child(3) { font-size: 1rem; margin: auto 0; }
  .user > span:nth-child(4) { font-size: 0.75rem; margin: auto 0; }

  .welcome > small {
    font-size: 0.75rem;
    font-weight: 300;
  }
  .welcome > p {
    font-size: 1rem;
    margin-bottom: 1rem;
  }

  .upgrade-banner {
    align-items: baseline;
    background: url(/assets/bnr_upgrade_full.jpg);
    background-repeat: no-repeat;
    background-size: contain;
    background-position: top;
    display: flex;
    flex-direction: column;
    height: 20rem;
    margin: -1rem -1rem 0;
    padding: 0 1rem 1.5rem;
    position: relative;
    text-align: left;
  }
  .upgrade-banner::after {
    background: url(/assets/img_wave_red.png);
    background-repeat: no-repeat;
    background-size: 100%;
    background-position: bottom;
    bottom: 0;
    content: '';
    height: 80%;
    left: 0;
    position: absolute;
    width: 100%;
  }
  .upgrade-banner > h2 {
    margin-top: auto;
    font-size: 1.5rem;
    z-index: 1;
  }
  .upgrade-banner > p {
    font-size: 1.25rem;
    z-index: 1;
  }
  .upgrade-banner > a {
    margin: 1rem auto 0;
    z-index: 1;
  }
}
