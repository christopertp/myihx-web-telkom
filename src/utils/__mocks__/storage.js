export const setToken = jest.fn();
export const getToken = jest.fn(() => true);
export const clearStorages = jest.fn();
export const setExpireTime = jest.fn();
export const checkExpireTime = jest.fn(() => false);
export const setUserData = jest.fn();
export const getUserData = jest.fn(() => ({ profilePicture: 'tes' }));
export const setGuestAddress = jest.fn();
export const getGuestAddress = jest.fn(() => '');
export const clearGuestAddress = jest.fn();
