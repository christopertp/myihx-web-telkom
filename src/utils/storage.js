const EXPIRE_TIME_STORAGE = 'myihx_web_expire_time';
const GUEST_ADDRESS_STORAGE = 'myihx_web_guest_address';
const SESSION_TIME = 'myihx_web_session_time';
const TOKEN_STORAGE = 'myihx_web_access_token';
const USER_DATA_STORAGE = 'myihx_web_user_data';

export function setToken(value) {
  localStorage.setItem(TOKEN_STORAGE, value);
}

export function getToken() {
  return localStorage.getItem(TOKEN_STORAGE);
}

export function clearStorages() {
  localStorage.removeItem(TOKEN_STORAGE);
  localStorage.removeItem(EXPIRE_TIME_STORAGE);
  localStorage.removeItem(USER_DATA_STORAGE);
  localStorage.removeItem(GUEST_ADDRESS_STORAGE);
  localStorage.removeItem(SESSION_TIME);
}

export function setExpireTime(value) {
  localStorage.setItem(EXPIRE_TIME_STORAGE, value * 1000);
}

export function checkExpireTime() {
  const time = new Date().getTime();
  const expire = localStorage.getItem(EXPIRE_TIME_STORAGE) || 0;

  return time > expire;
}

export function setUserData(value) {
  localStorage.setItem(USER_DATA_STORAGE, JSON.stringify(value));
}

export function getUserData() {
  const retval = localStorage.getItem(USER_DATA_STORAGE);

  return JSON.parse(retval) || '';
}

export function setGuestAddress(value) {
  localStorage.setItem(GUEST_ADDRESS_STORAGE, JSON.stringify(value));
}

export function getGuestAddress() {
  const retval = localStorage.getItem(GUEST_ADDRESS_STORAGE);

  return JSON.parse(retval) || '';
}

export function clearGuestAddress() {
  localStorage.removeItem(GUEST_ADDRESS_STORAGE);
}

export function setSessionTime(value) {
  localStorage.setItem(SESSION_TIME, value);
}

export function getSessionTime() {
  const retval = localStorage.getItem(SESSION_TIME);

  return parseInt(retval) || 0;
}
