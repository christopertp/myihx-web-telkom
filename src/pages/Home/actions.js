import { billingInfo } from '../../services/billing';
import { notifCard } from '../../services/payment';
import { home } from '../../services/product';
import { profile, statusCard } from '../../services/user';
import { BILL_FETCHED, CARD_FETCHED, DATA_FETCHED, LOADING, PROFILE_FETCHED, NOTIF_FETCHED } from './constants';

export function fetchBill() {
  return async dispatch => {
    dispatch(loadingAction(true, 'b'));
    try {
      const { data } = await billingInfo();
      dispatch(billingFetchedAction(data));
    } catch {
      dispatch(billingFetchedAction(''));
    }
  };
}

export function fetchHome() {
  return async dispatch => {
    dispatch(loadingAction(true, 'h'));
    try {
      const { data } = await home();
      const { latestNews, latestOffers, subscriptions } = data;
      dispatch(dataFetchedAction(latestNews, latestOffers, subscriptions));
    } catch (err) {
      dispatch(loadingAction(false, 'h'));
    }
  };
}

export function fetchProfile() {
  return async dispatch => {
    dispatch(loadingAction(true, 'p'));
    try {
      const { data } = await profile();
      const { fullName, profilePicture } = data;
      dispatch(profileFetchedAction(fullName, profilePicture));
    } catch (err) {
      dispatch(loadingAction(false, 'p'));
    }
  };
}

export function fetchNotifCard() {
  return async dispatch => {
    dispatch(loadingAction(true, 'n'));
    try {
      const { data } = await notifCard();
      dispatch(notifFetchedAction(data));
    } catch (err) {
      dispatch(loadingAction(false, 'n'));
    }
  };
}

export function fetchStatusCard() {
  return async dispatch => {
    dispatch(loadingAction(true, 'c'));
    try {
      const { data } = await statusCard();
      dispatch(cardFetchedAction(data));
    } catch (err) {
      dispatch(loadingAction(false, 'c'));
    }
  };
}

function loadingAction(isLoading, key) {
  return { type: LOADING, isLoading, key };
}

function billingFetchedAction(bill) {
  return { type: BILL_FETCHED, bill };
}

function cardFetchedAction(card) {
  return { type: CARD_FETCHED, card };
}

function dataFetchedAction(latestNews, latestOffers, subscriptions) {
  return { type: DATA_FETCHED, latestNews, latestOffers, subscriptions };
}

function notifFetchedAction(notif) {
  return { type: NOTIF_FETCHED, notif };
}

function profileFetchedAction(fullName, profilePicture) {
  return { type: PROFILE_FETCHED, fullName, profilePicture };
}
