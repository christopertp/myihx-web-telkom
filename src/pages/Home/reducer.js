import { BILL_FETCHED, CARD_FETCHED, DATA_FETCHED, LOADING, NOTIF_FETCHED, PROFILE_FETCHED } from './constants';

const initialState = {
  bill: '',
  card: {},
  fullName: '',
  isLoading: { b: false, c: true, h: true, n: true, p: true },
  latestNews: [],
  latestOffers: [],
  notif: {},
  profilePicture: '',
  subscriptions: '',
};

export default function reducer(state = initialState, action = {}) {
  const { type, bill, card, fullName, isLoading, key,
    latestNews, latestOffers, profilePicture, subscriptions, notif } = action;

  switch (type) {
    case BILL_FETCHED:
      return {
        ...state,
        bill,
        isLoading: { ...state.isLoading, b: false },
      };
    case CARD_FETCHED:
      return {
        ...state,
        card,
        isLoading: { ...state.isLoading, c: false },
      };
    case DATA_FETCHED:
      return {
        ...state,
        isLoading: { ...state.isLoading, h: false },
        latestNews,
        latestOffers,
        subscriptions,
      };
    case LOADING:
      return {
        ...state,
        isLoading: { ...state.isLoading, [key]: isLoading },
      };
    case NOTIF_FETCHED:
      return {
        ...state,
        notif,
        isLoading: { ...state.isLoading, n: false },
      };
    case PROFILE_FETCHED:
      return {
        ...state,
        fullName,
        isLoading: { ...state.isLoading, p: false },
        profilePicture,
      };
    default:
      return state;
  }
}
