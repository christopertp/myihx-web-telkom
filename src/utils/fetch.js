import axios from 'axios';
import { getToken } from './storage';

const lng = 'en';
export const BASIC_AUTH = {
  'Authorization': 'Basic dGVsa29tOmRhMWMyNWQ4LTM3YzgtNDFiMS1hZmUyLTQyZGQ0ODI1YmZlYQ==',
  'Accept-Language': lng,
};
export const BEARER_AUTH = {
  'Authorization': `Bearer ${getToken()}`,
  'Accept-Language': lng,
};

export default function fetch(url, method, param1, param2) {
  return new Promise((resolve, reject) => {
    axios[method](url, param1, param2)
      .then(res => resolve(res.data))
      .catch(err => {
        const defaultError = {
          code: 500,
          status: 'error',
          message: 'Failed to fetch data. Please contact developer.'
        };

        if (!err.response) {
          reject(defaultError);
        } else if (!err.response.data) {
          reject(defaultError);
        } else {
          reject(err.response.data);
        }
      });
  });
}

export function getUrl(urls) {
  const mode = process.env.MODE;

  if (mode === 'production') {
    return urls[0];
  }

  if (mode === 'staging') {
    return urls[1];
  }

  return urls[2];
}
