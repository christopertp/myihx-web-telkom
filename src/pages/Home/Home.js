import React, { createContext, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link, useLocation } from 'react-router-dom';
import ActivityCards from '../../components/elements/ActivityCards';
import Button from '../../components/elements/Button';
import PackageFinderLink from '../../components/elements/PackageFinderLink';
import Latest from '../../components/fragments/HomeLatest';
import NotificationCards from '../../components/elements/NotificationCards';
import PrimaryActions from '../../components/fragments/HomePrimaryActions';
import Recommended from '../../components/fragments/HomeRecommended';
import ShopMenu from '../../components/fragments/HomeShopMenu';
import Subscribe from '../../components/fragments/HomeSubscribe';
import { getToken } from '../../utils/storage';
import { fetchBill, fetchHome, fetchProfile, fetchStatusCard, fetchNotifCard } from './actions';
import styles from './styles.scoped.css';

const Context = createContext({});

export default function Home() {
  const { pathname } = useLocation();
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchHome());
  }, []);

  return (
    <Context.Provider value={{}}>
      {(pathname.match('/shop')) ? <Shop /> : <HomePage />}
    </Context.Provider>
  );
}

export function Shop() {
  return (
    <>
      <UpgradeBanner />
      <ShopMenu />
      <Recommended />
      <Latest />
      <Subscribe />
      <PackageFinderLink className={styles.finder} />
    </>
  );
}

export function UpgradeBanner() {
  return (
    <section className={styles['upgrade-banner']}>
      <h2>Have a need for speed?</h2>
      <p>Upgrade your speed for 10% off</p>
      <Button to="#" variant="white">Upgrade Package</Button>
    </section>
  );
}

export function HomePage() {
  const dispatch = useDispatch();
  const { card, isLoading, notif } = useSelector(s => s.home);

  useEffect(() => {
    if (getToken()) {
      dispatch(fetchBill());
      dispatch(fetchProfile());
      dispatch(fetchStatusCard());
      dispatch(fetchNotifCard());
    }
  }, []);

  return (
    <>
      <Welcome />
      <PrimaryActions />
      {!!getToken() && <NotificationCards isLoading={isLoading.n} notif={notif} />}
      {!!getToken() && <ActivityCards cards={card} heading="Upcoming" isLoading={isLoading.c} />}
      <Recommended />
      <Latest />
      <Subscribe />
    </>
  );
}

export function Welcome() {
  const { fullName, profilePicture } = useSelector(s => s.home);

  if (getToken()) {
    return (
      <Link className={styles.user} to="/profile">
        <figure style={{ backgroundImage: `url(${profilePicture})` }} />
        <span>Selamat Datang</span>
        <span>{fullName}</span>
        <span>0 pts</span>
      </Link>
    );
  }

  return (
    <section className={styles.welcome}>
      <small>Welcome to</small>
      <p>myIndiHome X</p>
    </section>
  );
}
