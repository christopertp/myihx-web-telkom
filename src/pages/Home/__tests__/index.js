import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { useLocation } from 'react-router-dom';
import ShallowRenderer from 'react-test-renderer/shallow';
import { getToken } from '../../../utils/storage';
import Home, { Welcome, HomePage, Shop, UpgradeBanner } from '../Home';

jest.mock('../../../utils/storage');

jest.mock('../actions', () => ({
  fetchBill: jest.fn(),
  fetchHome: jest.fn(),
  fetchProfile: jest.fn(),
  fetchStatusCard: jest.fn(),
  fetchNotifCard: jest.fn(),
}));

useSelector.mockImplementation(fn => {
  fn({});
  return { isLoading: {}, fullName: 'telkom', profilePicture: 'tes.jpg' };
});

describe('src/pages/Home', () => {
  test('render', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<Home />);
    expect(tree).toMatchSnapshot();
  });

  test('render Shop', () => {
    useLocation.mockImplementationOnce(() => ({ pathname: '/shop' }));
    const result = Home();
    expect(result.props.children.type).toBe(Shop);
  });

  test('HomePage', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<HomePage />);
    expect(tree).toMatchSnapshot();
  });

  test('HomePage guest', () => {
    getToken.mockImplementationOnce(() => false);

    const dispatch = jest.fn();
    useDispatch.mockImplementationOnce(() => dispatch);

    HomePage();

    expect(dispatch).toHaveBeenCalledTimes(0);
  });

  test('Welcome', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<Welcome />);
    expect(tree).toMatchSnapshot();

    getToken.mockImplementationOnce(() => false);
    const result = Welcome();
    expect(result.type).toBe('section');
    expect(result.props.children[0].props.children).toBe('Welcome to');
  });

  test('Shop', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<Shop />);
    expect(tree).toMatchSnapshot();
  });

  test('UpgradeBanner', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<UpgradeBanner />);
    expect(tree).toMatchSnapshot();
  });
});
