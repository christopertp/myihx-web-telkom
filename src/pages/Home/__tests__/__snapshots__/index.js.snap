// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/pages/Home HomePage 1`] = `
<React.Fragment>
  <Welcome />
  <HomePrimaryActions />
  <NotificationCards
    isLoading={false}
    notif={Object {}}
  />
  <ActivityCards
    cards={Object {}}
    fullWidth={false}
    heading="Upcoming"
    isLoading={false}
  />
  <HomeRecommended />
  <HomeLatest />
  <HomeSubscribe />
</React.Fragment>
`;

exports[`src/pages/Home Shop 1`] = `
<React.Fragment>
  <UpgradeBanner />
  <HomeShopMenu />
  <HomeRecommended />
  <HomeLatest />
  <HomeSubscribe />
  <PackageFinderLink
    className=""
  />
</React.Fragment>
`;

exports[`src/pages/Home UpgradeBanner 1`] = `
<section>
  <h2>
    Have a need for speed?
  </h2>
  <p>
    Upgrade your speed for 10% off
  </p>
  <Button
    className=""
    disabled={false}
    onClick={null}
    to="#"
    type="button"
    variant="white"
  >
    Upgrade Package
  </Button>
</section>
`;

exports[`src/pages/Home Welcome 1`] = `
<Link
  to="/profile"
>
  <figure
    style={
      Object {
        "backgroundImage": "url(tes.jpg)",
      }
    }
  />
  <span>
    Selamat Datang
  </span>
  <span>
    telkom
  </span>
  <span>
    0 pts
  </span>
</Link>
`;

exports[`src/pages/Home render 1`] = `
<Provider
  value={Object {}}
>
  <HomePage />
</Provider>
`;
